npm/networkPolicyController.go

```diff
@@ -0,0 +1,492 @@
+	if !ok {
+		tombstone, ok := obj.(cache.DeletedFinalStateUnknown)
+		if !ok {
+			metrics.SendErrorLogAndMetric(util.NetpolID, "[NETPOL DELETE EVENT] Received unexpected object type: %v", obj)
+			utilruntime.HandleError(fmt.Errorf("error decoding object, invalid type"))
+			return
+		}
+
+		if _, ok = tombstone.Obj.(*networkingv1.NetworkPolicy); !ok {
+			metrics.SendErrorLogAndMetric(util.NetpolID, "[NETPOL DELETE EVENT] Received unexpected object type (error decoding object tombstone, invalid type): %v", obj)
+			utilruntime.HandleError(fmt.Errorf("error decoding object tombstone, invalid type"))
+			return
+		}
+	}
+
+	var key string
+	var err error
+	if key, err = cache.MetaNamespaceKeyFunc(obj); err != nil {
+		utilruntime.HandleError(err)
+		return
+	}
+
+	netPolCachedKey := util.GetNSNameWithPrefix(key)
+
+	// (TODO): need to decouple this lock from npMgr if possible
+	c.npMgr.Lock()
+	_, netPolExists := c.npMgr.RawNpMap[netPolCachedKey]
+	c.npMgr.Unlock()
+	// If a network policy object is not in the RawNpMap, do not need to clean-up states for the network policy
+	// since netPolController did not apply for any states for the network policy
+	if !netPolExists {
+		return
+	}
+
+	c.workqueue.Add(key)
+}
+
+func (c *networkPolicyController) Run(threadiness int, stopCh <-chan struct{}) error {
+	defer utilruntime.HandleCrash()
+	defer c.workqueue.ShutDown()
+
+	klog.Infof("Starting Network Policy %d worker(s)", threadiness)
+	for i := 0; i < threadiness; i++ {
+		go wait.Until(c.runWorker, time.Second, stopCh)
+	}
+
+	klog.Infof("Started Network Policy %d worker(s)", threadiness)
+	<-stopCh
+	klog.Info("Shutting down Network Policy workers")
+
+	return nil
+}
+
+func (c *networkPolicyController) runWorker() {
+	for c.processNextWorkItem() {
+	}
+}
+
+func (c *networkPolicyController) processNextWorkItem() bool {
+	obj, shutdown := c.workqueue.Get()
+
+	if shutdown {
+		return false
+	}
+
+	err := func(obj interface{}) error {
+		defer c.workqueue.Done(obj)
+		var key string
+		var ok bool
+		if key, ok = obj.(string); !ok {
+			// As the item in the workqueue is actually invalid, we call
+			// Forget here else we'd go into a loop of attempting to
+			// process a work item that is invalid.
+			c.workqueue.Forget(obj)
+			utilruntime.HandleError(fmt.Errorf("expected string in workqueue but got %#v", obj))
+			return nil
+		}
+		// Run the syncNetPol, passing it the namespace/name string of the
+		// network policy resource to be synced.
+		// TODO : may consider using "c.queue.AddAfter(key, *requeueAfter)" according to error type later
+		if err := c.syncNetPol(key); err != nil {
+			// Put the item back on the workqueue to handle any transient errors.
+			c.workqueue.AddRateLimited(key)
+			return fmt.Errorf("error syncing '%s': %s, requeuing", key, err.Error())
+		}
+		// Finally, if no error occurs we Forget this item so it does not
+		// get queued again until another change happens.
+		c.workqueue.Forget(obj)
```

Comment:
How it is different from c.Workqueue.Done?

Reply:
// We call Done here so the workqueue knows we have finished processing this item. We also must remember to call Forget if we do not want this work item being re-queued. For example, we do not call Forget if a transient error occurs, instead the item is put back on the workqueue and attempted again after a back-off period. - Refer to 
https://github.com/kubernetes/sample-controller/blob/master/controller.go#L196